SPEC.md

```diff
@@ -77,33 +78,58 @@ Network configuration in JSON format is streamed to the plugin through stdin. Th
 
 ### Result
 
-Success is indicated by a return code of zero and the following JSON printed to stdout in the case of the ADD command. This should be the same output as was returned by the IPAM plugin (see [IP Allocation](#ip-allocation) for details).
+Success is indicated by a return code of zero and the following JSON printed to stdout in the case of the ADD command. The `ip` and `dns` items should be the same output as was returned by the IPAM plugin (see [IP Allocation](#ip-allocation) for details) except that the plugin should fill in the `interface` indexes appropriately, which are missing from IPAM plugin output since IPAM plugins should be unaware of interfaces.
 
 ```
 {
   "cniVersion": "0.1.0",
-  "ip4": {
-    "ip": <ipv4-and-subnet-in-CIDR>,
-    "gateway": <ipv4-of-the-gateway>,  (optional)
-    "routes": <list-of-ipv4-routes>    (optional)
-  },
-  "ip6": {
-    "ip": <ipv6-and-subnet-in-CIDR>,
-    "gateway": <ipv6-of-the-gateway>,  (optional)
-    "routes": <list-of-ipv6-routes>    (optional)
-  },
+  "interfaces": [
```

Comment:
@eyakubovich @rosenhouse the use-cases that I currently have in mind are:

1. host port<->container port mapping which, if you're using Docker networking you get for free but of course this is CNI so you have to invent it yourself.  We've done that in Kubernetes but the iptables rules normally used for this require the host-side bridge interface name.  I suppose we could write a CNI port forwarding plugin and chain it but that would still require the host bridge interface name.
2. traffic shaping which typically requires the host bridge interface name and the host-side veth interface name.  Could be done as a CNI plugin or utility library too, but still needs chaining.
3. future VM/hypervisor based runtimes, where the networking setup needs to be done before the VM actually starts, so the thing starting the VM needs to know about the host-side resources the CNI plugin set up.  For example, after the CNI plugin creates a macvtap interface the VM-based runtime would then find the associated character device and pass that to qemu.

Maybe @steveeJ can comment more too, since he and I worked together to develop this specific API proposal a few weeks ago.